### ISSUE TITLE
WARNING: excessive pending puts (> 16384), closing stream

### ISSUE THREAD

**Issue by lkrubner** [ztellman/manifold#46]
I have a very simple app (less than 200 lines of code) and I thought Manifold would be a good way to set up a simple jobs queue, but when I run the app, after about 15 to 30 minutes, I get this error:

     Aug 10, 2015 11:37:19 PM clojure.tools.logging$eval36$fn__40 invoke
     WARNING: excessive pending puts (> 16384), closing stream
     java.lang.IllegalStateException
	at manifold.stream.default.Stream.put(default.clj:111)
	at manifold.stream.default.Stream.put(default.clj:139)
	at eh.streams$source_worker$fn__141.invoke(streams.clj:58)
	at eh.streams$source_worker.invoke(streams.clj:55)

I have a "start" function that spins up 5 "source" workers and 5 "destination" workers:

     (defn start
        [map-of-config-info]
        (dotimes [_ 5]
         (future (source-worker map-of-config-info))
         (future (destination-worker))))

The source-worker pulls records out of Redis and puts them in an anonymous function for some processing and then puts the function on a stream. The destination-worker simply pulls the functions off the stream and executes them: 

     (defn destination-worker []
       (loop [event-fn (stream/take! event-stream)]
         (@event-fn)
         (recur (stream/take! event-stream))))

I'm guessing the destination-workers run slowly, because they do all the work, whereas the source-workers do very little work, and so I would guess that slowly, over the course of 20 minutes or so, the puts! build up in excess of the takes!, and so I get this error? How could I test that? 

I will experiment with having more destination-workers and less source-workers, though that seems a bit inexact.

**Comment by ztellman** [comment]
Sorry, missed this somehow.  I don't see in your code where you're calling `put!`, but that's where your problem is.  You should be dereferencing the value returned by `put!`, which will allow the enqueuers to wait if the downstream workers are not processing fast enough.

**Comment by lkrubner** [comment]
Am I reading you correctly? " the value returned by put!" I must have misunderstood how to use the library. (Possibly I carried over an inaccurate idea from Lamina.). I throw away the return value of put!:

      (defn enqueue
        "2015-07-26 - this is an unusual situation where I deliberately return nil so that anyone
         calling enqueue is aware that they must not expect a return value from that function.
         Rather, if they want to get data from the database, they must pass in a deferred, and wait
         till a value is delivered to that deferred."
        [message eventual-result]
        {:pre [(map? message)
               (do (println  (type eventual-result)) true)]
         :post [(nil? %)]}
        (stream/put! @message-stream
          [(fn [db-connection] (persistence/make-consistent message db-connection))
           eventual-result])
          nil)

@message-stream is an atom with a stream inside of it. 

"eventual-result" is a deferred. Somewhere on the outside there is code waiting to deref the value of eventual-result. Here it gets passed along. 

A function (a closure) gets put on the stream. There are elsewhere workers that are pulling these closures off of the stream and executing them, thus running a query against the database, with the result being stored in eventual-result. And, as I said, somewhere there is some code waiting for a value to be delivered to eventual-result. 

But you say I should actually do something with the return value of put!

I eventually got around the problem by having a lot more workers pull the closures off the queue. But that feels like an imprecise hack. I would like to know the correct way to do this.

**Comment by ztellman** [comment]
In a default stream, without a buffer, a `put!` won't succeed without a matching `take!`.  This allows you to make sure the rate of production doesn't outstrip the rate of consumption, which is what's happening here.  You're calling `enqueue` repeatedly without checking to see if you already have too many waiting (where "too many" in this case is more than 0, since you haven't defined a buffer).

The correct way to do this is wait on the result of `put!`, and possibly define a buffer size so that you can have a few tasks waiting for the consumers at all times.  A more in-depth explanation about backpressure and queues can be found at https://www.youtube.com/watch?v=1bNOO3xxMc0.

**Comment by lkrubner** [comment]
Thank you for that explanation. I'd already seen your video about backpressure, but I did not think about how it applied here. 
I'll add a buffer to better manage the situation.
